tools/mobimeta.1.in
.Dd February 16, 2022
.Dt mobimeta 1 URM
.Os Unix
.Sh NAME
.Nm mobimeta
.Nd Utility for modifying metadata of MOBI format ebook files.
.Sh SYNOPSIS
.Nm
.Op Fl a Ar meta=value Ns Op Ar ,meta=value,...
.Op Fl d Ar meta Ns Op Ar ,meta,...
.Op Fl s Ar meta=value Ns Op Ar ,meta=value,...
.if !'@ENCRYPTION_OPT@'yes' .ig
.Op Fl p Ar pid
.Op Fl P Ar serial
..
.Op Fl hv
.Ar filein
.Op Ar fileout
.Sh DESCRIPTION
The program handles .prc, .mobi, .azw; .azw3, .azw4, some .pdb documents. Written as a test case for
.Nm libmobi
library.
.Pp
Invoked without arguments prints document metadata and exits. If
.Ar fileout
is specified, modified document will be saved with this name, otherwise original input
.Ar filein
will be modified.
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent
.It Fl a Ar meta=value Ns Op Ar ,meta=value,...
add (append) metadata
.It Fl d Ar meta Ns Op Ar ,meta,...
delete metadata
.It Fl s Ar meta=value Ns Op Ar ,meta=value,...
set (replace) metadata
.It Ar meta
for a list of valid named meta keys try
.Nm mobimeta Bro Fl a | Fl d | Fl s Brc
.Ar \&? .
If meta is an integer it will be treated as a numeric EXTH record key (expert usage).
.It Ar value
new value that will be set for a given meta key
.if !'@ENCRYPTION_OPT@'yes' .ig
.It Fl p Ar pid
set pid for decryption
.It Fl P Ar serial
set device serial number for decryption
..
.It Fl h
show usage summary and exit
.It Fl v
show version and exit
.El
.Pp
.Sh EXAMPLES
The following command will set new title, new author and will delete all publisher metadata (if any).
.Pp
.Bd -filled -offset indent
.Nm mobimeta
.Fl s
title="My title",author="My name"
.Fl d
publisher example.mobi
.Ed
.Pp
The following command appends EXTH record identified by its numeric key 204 with numeric value 201.
.Pp
.Bd -filled -offset indent
.Nm mobimeta
.Fl a
204=201 example.mobi
.Ed
.Pp
The following command will list valid meta named keys.
.Pp
.Bd -filled -offset indent
.Nm mobimeta
.Fl a
?
.Ed
.Sh RETURN VALUES
The
.Nm mobimeta
utility returns 0 on success, 1 on error.
.Sh COPYRIGHT
Copyright (C) 2014-2022 Bartek Fabiszewski.
.Pp
Released under LGPL version 3 or any later (same as libmobi).
.Sh WEB SITE
Visit http://www.fabiszewski.net for details.
.Sh DIAGNOSTICS
For diagnostics
.Nm libmobi
must be configured with
.Op Fl Fl enable-debug
option.
.Sh SEE ALSO
.Xr mobitool 1


ChangeLog
2021-10-20: Version 0.9
2021-10-24: Fix out-of-tree build
2021-10-22: Fix mingw build, code formatting
2021-10-14: Fix gcc format truncation warning
2021-10-14: Include autogen.sh in distribution bundle
2021-10-14: Create codeql-analysis.yml
2021-10-14: Fix autoconf 2.70 warnings, clean up
2021-10-14: Build fails with autoconf 2.70
2021-10-11: Version 0.8
2021-10-11: Update Xcode project
2021-10-11: Fix warnings about changed signedness
2021-09-18: Fix potential out-of-buffer read while parsing corrupt file, closes #38
2021-09-18: Fix potential out-of-buffer read while parsing corrupt file, closes #35, #36
2021-09-09: Version 0.7
2021-09-09: Merge pull request #33 from occia/public
2021-09-09: fix oob write bug inside libmobi
2021-06-07: Merge pull request #30 from codetheweb/patch-1
2021-06-07: Add reference to brew formula
2020-09-02: Fix null pointer dereference in case of broken fragment
2020-08-01: Update changelog
2020-08-01: Version 0.6
2020-07-31: Fix typo
2020-07-31: Add Readme to dist package
2020-07-31: Remove anchor on truncated link
2020-07-31: Fix missing option in man page
2020-07-30: Include test samples in dist package
2020-07-25: Fix gcc 7+ warnings about implicit fall through and format truncation
2020-07-24: Unique names for internal functions to avoid confilicts with static linking
2020-06-24: Close file in error branch
2020-06-24: Fix static compilation with miniz on gcc
2020-06-24: Minor documentation fixes
2020-06-23: Version 0.5
2020-06-23: mobitool: add dump cover option
2020-06-23: Minor documentation improvement
2020-06-23: Fix potential buffer over-read
2019-03-18: Fix: try also "name" attribute when searching for link anchor tags, closes #24
2019-02-22: Add mobi_is_replica function
2019-02-22: Fix potential read beyond buffer
2019-02-22: Travis migration
2018-08-07: Fix: missing items in recreated ncx file
2018-06-20: Fix: printf format warning on some gcc versions
2018-06-20: Fix: make dist broken by nonexistent header files
2018-06-20: VERSION 0.4
2018-06-20: Fix: buffer overflow (CVE-2018-11726)
2018-06-20: Fix: buffer overflow (CVE-2018-11724)
2018-06-20: Fix: read beyond buffer (CVE-2018-11725)
2018-06-20: Fix: buffer overflow (mobitool), closes #18
2018-06-20: Fix: read beyond buffer with corrupted KF8 Boundary record, closes #19
2018-06-20: Fix: read beyond buffer, closes #16, #17
2018-06-20: Updated xcode project files
2018-04-03: Fix: ncx part was not scanned for links, fixes #12
2018-04-02: Fix regression, potential use after free
2018-04-02: Skip broken resources, fixes #10
2018-03-05: Allow processing zero length text records, fixes #9
2017-12-25: Skip broken first resource offset instead of dying
2017-12-18: Skip broken links reconstruction instead of dying
2017-11-27: Disable travis OS X builds, as they usually time out
2017-11-16: Fix: increase max number of dictionary entries per record
2017-11-14: Fix for some encrypted documents with palmdoc encoding
2017-11-06: Fix: potential null pointer dereference
2017-10-16: Manpage cleanup
2017-09-27: Update README
2017-09-26: Increase maximum length of attribute name and value, closes #5
2017-02-26: Remove obsolete files from VS build (closes #3) [ci skip]
2016-11-05: Mobitool: use epub extension if extracted source resource is epub
2016-06-10: Update docs
2016-06-10: Update test files
2016-06-10: Fix: out of bounds read in corrupt font resource
2016-06-10: Prevent memory leak in case of corrupt font resources
2016-06-10: Calculate deobfuscation buffer limit from key length
2016-06-10: Merge branch 'public' of ssh://fabiszewski.net/home/git/libmobi into public
2016-06-10: Fix: USE_LIBXML2 macro was not included from config.h
2016-06-10: Fix: USE_LIBXML2 macro was not included from config.h
2016-06-09: Fix: memory leak in tools
2016-06-09: Fix: potential out of bounds read
2016-06-09: Fix: memory leak in internal xmlwriter
2016-06-01: Update README
2016-05-19: Feature: verify decryption key type
2016-05-19: Cleanup converting little endian buffer to 32-bit integer
2016-05-19: Feature: check drm expiration dates
2016-05-18: Fix: memory leaks in encryption
2016-05-18: Fix concurrent autotools builds
2016-05-18: use relative path, as $(top_srcdir) fails to be substituted (?)
2016-05-18: update vcxproj
2016-05-18: Include headers in automake sources
2016-05-18: Fix: automake out-of-tree miniz build
2016-05-18: Fix: wrongly detected fdst record broke some ancient documents
2016-05-18: Fix: improve index header parsing, some old dictionaries might not load
2016-05-18: Fix: convert encoding of opf strings from cp1252 indices
2016-05-18: Quiet warnings about unused values of wiped variables
2016-05-18: Fix: potential memory leak
2016-05-18: Fix: wrongly decoded "&copy;" entity
2016-05-16: Fix: huffdic decompression fails in case of huge documents
2016-05-14: Simplify buffer_init_null() function
2016-05-14: Use ARRAYSIZE macro
2016-05-14: Feature: calculate pid for decryption from device serial number
2016-04-29: Use endian-independent byte swapping
2016-04-29: Exclude unused miniz functions from binary
2016-04-29: Add SHA-1 routines
2016-04-27: Fix miniz.c formatting
2016-04-27: Documentation
2016-04-20: Update changelog
2016-04-20: Fix potential null pointer dereference
2016-04-20: Remove useless check
2016-04-20: Fix text record size calculation
2016-04-20: Fix buffer checking and freeing
2016-04-19: Update docs
2016-04-19: Update ChangeLog
2016-04-19: Fix comparison between signed and unsigned integer
2016-04-19: use strdup on linux/glibc
2016-04-19: Add initial write and metadata editing support. Add mobimeta tool.
2016-04-19: Always check whether memory allocation succeeded
2016-04-18: Fix: guarantee array resize step is at least 1
2016-04-13: Workaround to read some old mobipocket files
2016-04-13: Improve pdb dates resolving
2016-04-07: Minor documentation edit
2016-04-07: Update changelog
2016-04-06: Fix format warning
2016-04-06: Update test checksums

tools/mobimeta.c
/** @file mobimeta.c
 *
 * @brief mobimeta
 *
 * @example mobimeta.c
 * Program for testing libmobi library
 *
 * Copyright (c) 2020 Bartek Fabiszewski
 * http://www.fabiszewski.net
 *
 * Licensed under LGPL, either version 3, or any later.
 * See <http://www.gnu.org/licenses/>
 */

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>
#include <string.h>
#include <mobi.h>

#include "common.h"

/* encryption */
#ifdef USE_ENCRYPTION
# define PRINT_ENC_USG " [-p pid] [-P serial]"
# define PRINT_ENC_ARG "p:P:"
#else
# define PRINT_ENC_USG ""
# define PRINT_ENC_ARG ""
#endif

/* command line options */
#ifdef USE_ENCRYPTION
bool setpid_opt = false;
bool setserial_opt = false;
#endif

/* options values */
#ifdef USE_ENCRYPTION
char *pid = NULL;
char *serial = NULL;
#endif

#define META_SIZE ARRAYSIZE(meta_functions)
#define ACTIONS_SIZE ARRAYSIZE(actions)

#if HAVE_ATTRIBUTE_NORETURN 
static void exit_with_usage(const char *progname) __attribute__((noreturn));
#else
static void exit_with_usage(const char *progname);
#endif

/**
 @brief Meta handling functions
 */
typedef MOBI_RET (*MetaFunAdd)(MOBIData *m, const char *string);
typedef MOBI_RET (*MetaFunDel)(MOBIData *m);

/**
 @brief Meta functions structure
 */
typedef struct {
    const char *name;
    MetaFunAdd function_add;
    MetaFunAdd function_set;
    MetaFunDel function_del;
} CB;

const CB meta_functions[] = {
    { "author", mobi_meta_add_author, mobi_meta_set_author, mobi_meta_delete_author },
    { "title", mobi_meta_add_title, mobi_meta_set_title, mobi_meta_delete_title },
    { "publisher", mobi_meta_add_publisher, mobi_meta_set_publisher, mobi_meta_delete_publisher },
    { "imprint", mobi_meta_add_imprint, mobi_meta_set_imprint, mobi_meta_delete_imprint },
    { "description", mobi_meta_add_description, mobi_meta_set_description, mobi_meta_delete_description },
    { "isbn", mobi_meta_add_isbn, mobi_meta_set_isbn, mobi_meta_delete_isbn },
    { "subject", mobi_meta_add_subject, mobi_meta_set_subject, mobi_meta_delete_subject },
    { "publishdate", mobi_meta_add_publishdate, mobi_meta_set_publishdate, mobi_meta_delete_publishdate },
    { "review", mobi_meta_add_review, mobi_meta_set_review, mobi_meta_delete_review },
    { "contributor", mobi_meta_add_contributor, mobi_meta_set_contributor, mobi_meta_delete_contributor },
    { "copyright", mobi_meta_add_copyright, mobi_meta_set_copyright, mobi_meta_delete_copyright },
    { "asin", mobi_meta_add_asin, mobi_meta_set_asin, mobi_meta_delete_asin },
    { "language", mobi_meta_add_language, mobi_meta_set_language, mobi_meta_delete_language },
};

/**
 @brief Print usage info
 @param[in] progname Executed program name
 */
static void exit_with_usage(const char *progname) {
    char *p = strrchr(progname, separator);
    if (p) { progname = ++p; }
    printf("usage: %s [-a | -s meta=value[,meta=value,...]] [-d meta[,meta,...]]" PRINT_ENC_USG " [-hv] filein [fileout]\n", progname);
    printf("       without arguments prints document metadata and exits\n");
    printf("       -a ?           list valid meta named keys\n");
    printf("       -a meta=value  add metadata\n");
    printf("       -d meta        delete metadata\n");
    printf("       -s meta=value  set metadata\n");
#ifdef USE_ENCRYPTION
    printf("       -p pid         set pid for decryption\n");
    printf("       -P serial      set device serial for decryption\n");
#endif
    printf("       -h             show this usage summary and exit\n");
    printf("       -v             show version and exit\n");
    exit(ERROR);
}

/**
 @brief Check whether string is integer
 @param[in] string String
 @return True if string represents integer
 */
static bool isinteger(const char *string) {
    if (*string == '\0') { return false; }
    while (*string) {
        if (!isdigit(*string++)) { return false; }
    }
    return true;
}

/**
 @brief Parse suboption's list key=value[,key=value,...]
 and get first key-value pair.
 @param[in,out] subopts List of suboptions
 @param[in,out] token Will be filled with first found key name or NULL if missing
 @param[in,out] value Will be filled with first found key value or NULL if missing
 @return True if there are more pairs to parse
 */
static bool parsesubopt(char **subopts, char **token, char **value) {
    if (!**subopts) { return false; }
    *token = NULL;
    *value = NULL;
    char *p = NULL;
    while ((p = (*subopts)++) && *p) {
        if (*p == ',') {
            *p = '\0';
            return true;
        }
        if (!*token) { *token = p; }
        if (*p == '=') {
            *p = '\0';
            *value = ++p;
        }
    }
    return false;
}

/**
 @brief Get matching token from meta functions array
 @param[in] token Meta token name
 @return Index in array,-1 if not found
 */
static int get_meta(const char *token) {
    for (int i = 0; i < (int) META_SIZE; i++) {
        if (strcmp(token, meta_functions[i].name) == 0) {
            return i;
        }
    }
    return -1;
}

/**
 @brief Main
 
 @param[in] argc Arguments count
 @param[in] argv Arguments array
 @return SUCCESS (0) or ERROR (1)
 */
int main(int argc, char *argv[]) {
    if (argc < 2) {
        exit_with_usage(argv[0]);
    }
    
    typedef struct {
        int command;
        int meta;
        char *value;
    } Action;
    Action actions[100];
    
    size_t cmd_count = 0;
    
    char *token = NULL;
    char *value = NULL;
    char *subopts;
    int opt;
    int subopt;
    bool parse;
    while ((opt = getopt(argc, argv, "a:d:hs:" PRINT_ENC_ARG "v")) != -1) {
        switch (opt) {
            case 'a':
            case 'd':
            case 's':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", opt);
                    return ERROR;
                }
                subopts = optarg;
                parse = true;
                while (parse) {
                    parse = parsesubopt(&subopts, &token, &value);
                    if (token && (subopt = get_meta(token)) >= 0) {
                        if ((value == NULL || strlen(value) == 0) && opt != 'd') {
                            printf("Missing value for suboption '%s'\n\n", meta_functions[subopt].name);
                            exit_with_usage(argv[0]);
                        }
                        if (cmd_count < (ACTIONS_SIZE)) {
                            actions[cmd_count++] = (Action) { opt, subopt, value };
                        }
                    } else if (token && isinteger(token)) {
                        if ((value == NULL || strlen(value) == 0) && opt != 'd') {
                            printf("Missing value for suboption '%s'\n\n", token);
                            exit_with_usage(argv[0]);
                        }
                        if (cmd_count < (ACTIONS_SIZE)) {
                            /* mark numeric meta with upper opt */
                            actions[cmd_count++] = (Action) { toupper(opt), atoi(token), value };
                        }
                    } else {
                        if (token == NULL || token[0] != '?') {
                            printf("Unknown meta: %s\n", token ? token : optarg);
                        }
                        printf("Valid named meta keys:\n");
                        for (size_t i = 0; i < META_SIZE; i++) {
                            printf("   %s\n", meta_functions[i].name);
                        }
                        printf("\n");
                        if (token == NULL || token[0] != '?') {
                            exit_with_usage(argv[0]);
                        }
                        return SUCCESS;
                    }
                }
                break;
#ifdef USE_ENCRYPTION
            case 'p':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", opt);
                    return ERROR;
                }
                setpid_opt = true;
                pid = optarg;
                break;
            case 'P':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", opt);
                    return ERROR;
                }
                setserial_opt = true;
                serial = optarg;
                break;
#endif
            case 'v':
                printf("mobimeta build: " __DATE__ " " __TIME__ " (" COMPILER ")\n");
                printf("libmobi: %s\n", mobi_version());
                return 0;
            case '?':
                if (isprint(optopt)) {
                    printf("Unknown option `-%c'\n", optopt);
                }
                else {
                    printf("Unknown option character `\\x%x'\n", optopt);
                }
                exit_with_usage(argv[0]);
            case 'h':
            default:
                exit_with_usage(argv[0]);
        }
    }
    
    int file_args = argc - optind;
    if (file_args <= 0) {
        printf("Missing filename\n");
        exit_with_usage(argv[0]);
    }
    char infile[FILENAME_MAX];
    strncpy(infile, argv[optind], FILENAME_MAX - 1);
	infile[FILENAME_MAX - 1] = '\0';
    
    if (file_args >= 2) { optind++; }
    
    char outfile[FILENAME_MAX];
    strncpy(outfile, argv[optind], FILENAME_MAX - 1);
	outfile[FILENAME_MAX - 1] = '\0';
    
    /* Initialize MOBIData structure */
    MOBIData *m = mobi_init();
    if (m == NULL) {
        printf("Libmobi initialization failed\n");
        return ERROR;
    }
    
    /* read */
    FILE *file_in = fopen(infile, "rb");
    if (file_in == NULL) {
        mobi_free(m);
        int errsv = errno;
        printf("Error opening file: %s (%s)\n", infile, strerror(errsv));
        return ERROR;
    }
    MOBI_RET mobi_ret = mobi_load_file(m, file_in);
    fclose(file_in);
    if (mobi_ret != MOBI_SUCCESS) {
        mobi_free(m);
        printf("Error loading file (%s)\n", libmobi_msg(mobi_ret));
        return ERROR;
    }
    
#ifdef USE_ENCRYPTION
    if (setpid_opt || setserial_opt) {
        int ret = set_decryption_key(m, serial, pid);
        if (ret != SUCCESS) {
            mobi_free(m);
            return ret;
        }
    }
#endif
    
    if (cmd_count == 0) {
        print_summary(m);
        print_exth(m);
        mobi_free(m);
        return SUCCESS;
    }
    
    for (size_t i = 0; i < cmd_count; i++) {
        Action action = actions[i];
        MOBIExthMeta tag;
        mobi_ret = MOBI_SUCCESS;
        switch (action.command) {
                /* named meta */
            case 'a':
                mobi_ret = meta_functions[action.meta].function_add(m, action.value);
                break;
            case 'd':
                mobi_ret = meta_functions[action.meta].function_del(m);
                break;
            case 's':
                mobi_ret = meta_functions[action.meta].function_set(m, action.value);
                break;
                /* numeric exth */
            case 'A':
                tag = mobi_get_exthtagmeta_by_tag((MOBIExthTag) action.meta);
                if (tag.name && tag.type == EXTH_NUMERIC && isinteger(action.value)) {
                    uint32_t numeric = (uint32_t) atoi(action.value);
                    mobi_ret = mobi_add_exthrecord(m, (MOBIExthTag) action.meta, sizeof(uint32_t), &numeric);
                } else {
                    mobi_ret = mobi_add_exthrecord(m, (MOBIExthTag) action.meta, (uint32_t) strlen(action.value), action.value);
                }
                break;
            case 'D':
                mobi_ret = mobi_delete_exthrecord_by_tag(m, (MOBIExthTag) action.meta);
                break;
            case 'S':
                mobi_ret = mobi_delete_exthrecord_by_tag(m, (MOBIExthTag) action.meta);
                if (mobi_ret != MOBI_SUCCESS) { break; }
                tag = mobi_get_exthtagmeta_by_tag((MOBIExthTag) action.meta);
                if (tag.name && tag.type == EXTH_NUMERIC && isinteger(action.value)) {
                    uint32_t numeric = (uint32_t) atoi(action.value);
                    mobi_ret = mobi_add_exthrecord(m, (MOBIExthTag) action.meta, sizeof(uint32_t), &numeric);
                } else {
                    mobi_ret = mobi_add_exthrecord(m, (MOBIExthTag) action.meta, (uint32_t) strlen(action.value), action.value);
                }
                break;
        }
        if (mobi_ret != MOBI_SUCCESS) {
            printf("Metadata modification failed (%s)\n", libmobi_msg(mobi_ret));
            mobi_free(m);
            return ERROR;
        }
        
    }
    
    /* write */
    printf("Saving %s...\n", outfile);
    FILE *file_out = fopen(outfile, "wb");
    if (file_out == NULL) {
        mobi_free(m);
        int errsv = errno;
        printf("Error opening file: %s (%s)\n", outfile, strerror(errsv));
        return ERROR;
    }
    mobi_ret = mobi_write_file(file_out, m);
    fclose(file_out);
    if (mobi_ret != MOBI_SUCCESS) {
        mobi_free(m);
        printf("Error writing file (%s)\n", libmobi_msg(mobi_ret));
        return ERROR;
    }
    
    /* Free MOBIData structure */
    mobi_free(m);
    
    return SUCCESS;
}
